prep/go/tutorial/chapter4/xkcd.go

```diff
@@ -0,0 +1,392 @@
+    err = os.Mkdir(storageDir, 0777)
+    if err != nil && !os.IsExist(err) {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+    storageDirFmt := strings.Join([]string{storageDir, "%d"}, "")
+    return storageDirFmt
+}
+
+func main() {
+    XKCD_URL_FMT = "https://xkcd.com/%d/"
+    XKCD_URL_JSON_FMT = "https://xkcd.com/%d/info.0.json"
+    COMIC_STORAGE_FMT = mkComicStorageDir()
+    homeDir, err := os.UserHomeDir()
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+    COMIC_INDEX_STORAGE = homeDir + "/.local/xkcd_cmd/index"
+
+    if len(os.Args) < 2 {
+        fmt.Printf("Provide a comic by id or search terms\n")
+        os.Exit(1)
+    }
+
+    comicId, err := strconv.Atoi(os.Args[1])
+    if comicId == 404 {
+        fmt.Printf("You should know better...\n")
+        os.Exit(1)
+    }
+
+    // In practice we know 3000 is a better starting capacity
+    // as xkcd has ~2790 comics but its nice to force
+    // the resize logic to run for testing + learning purposes
+    comics := make([]XKCDComic, 2000, 2000)
+    loadComics(&comics)
+
+    // TODO make id lookup fast (only load single comic from disk)
+    if err == nil && comicId > 0 {
+        // if requested by number print comic
+        if comicId > len(comics) {
+            printComics([]int{}, &comics)
+            os.Exit(0)
+        }
+        printComics([]int{comicId}, &comics)
+    } else {
+        index := loadAndBuildIndex(&comics)
+        ids := searchIndex(os.Args[1:], index)
+        printComics(ids, &comics)
+    }
+}
+
+func printComics(ids []int, comics *[]XKCDComic) {
+    comicViews := make([]XKCDComicView, 0, 0)
+
+    for _, v := range ids {
+        view := toView((*comics)[v])
+        comicViews = append(comicViews, view)
+    }
+
+    data, err := json.MarshalIndent(comicViews, "", "  ")
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+    fmt.Println(string(data))
+}
+
+type WeightedId struct {
+    Id int
+    Weight int
+}
+
+func min(a, b int) int {
+    if a < b {
+        return a
+    }
+    return b
+}
+
+// Case insensitive search the term-based inverted index by matching exact terms
+// The more terms that match the higher the weight
+// For example "Three Laws of Robotics" will return the comics that
+// Mention that phrase at the top of the list.
+// Hardcode results to 3 since it can be comfortable viewed
+// TODO make flag for result limit
+// TODO (Algorithms) could use levelstien distance to search on typo'd terms
+func searchIndex(terms []string, index ComicIndex) []int {
+    weightedIds := make([]WeightedId, 0, 0)
+    idsMapToWeight := make(map[int]int)
+    ids := make([]int, 0, 3)
+
+    for _, term := range terms {
+        for comicId, _ := range index[strings.ToLower(term)] {
+            // TODO I did term counting here to do this quickly but I should store
+            // the term counts in the index itself instead of calculating at search time
```

Comment:
Wrong. I'm weighting on how many terms are in the document which can only be calculated at search time.

I can still add term frequency to index but that doesn't invalidate the counting here.